### ISSUE TITLE
[Question] About the ProportionalBatchSampler 

### ISSUE THREAD

**Issue by claeyzre** [UKPLab/sentence-transformers#3040]
Hello @tomaarsen,

Thank you very much for your work on this lib.

I am trying to reproduce [nomic-ai contrastors](https://arxiv.org/pdf/2402.01613) training using another model as base and the SentenceTransformerTrainer.
During the contrastive-pretraining phase they use the `InfoNCE `loss with a very large batch size, using grad cache. I assume it's exactly similar to using the `CachedMultipleNegativeRankingLoss`.

They emphasize on the following:

> During training, we sample pairs from one data source at a time and fill the entire batch with samples from that single source to discourage the model from learning source-specific shortcuts.

From my understanding the `ProportionalBatchSampler `does that by nature as it samples batch one dataset at a time w.r.t its proportion in the mix, but it does not mix the dataset samples within one batch right ? 

Thank you very much,

**Comment by tomaarsen** [comment]
Hello!

Indeed, InfoNCE is like `MultipleNegativesRankingLoss`, and InfoNCE with GradCache is equivalent to `CachedMultipleNegativeRankingLoss`.

Both of the multi dataset samplers only sample texts from the same dataset within each batch. In Sentence Transformers, this was more of a design decision because each training/evaluation dataset can have 1) different columns and 2) different losses. One batch with different columns, passed to different losses, just didn't make sense. So, RoundRobin or Proportional both correspond with your quote of filling each batch with samples from one source.

There's a bit of documentation on the samplers here: https://sbert.net/docs/package_reference/sentence_transformer/sampler.html#multidatasetbatchsamplers
In short:
* `MultiDatasetBatchSamplers.ROUND_ROBIN`: Cycle through all training datasets until one is empty, then stop. So, all samples from the smallest dataset are used, and the same amount of samples are used from all other datasets.
* `MultiDatasetBatchSamplers.PROPORTIONAL`: Sample each batch from each dataset randomly. So, all samples from all datasets are used: larger datasets are sampled from more often.

I hope this clears it up a bit!

- Tom Aarsen

**Comment by claeyzre** [comment]
That's pretty much what I had in mind ! Thank you very much
